### PR TITLE
Webhooks upgrade complete fix

### LINKED ISSUE #1279: Issue with NIOHTTPClientUpgradeHandler when immediatly firing write after buffered read.
In [NIOHTTPClientUpgradeHandler.swift](https://github.com/apple/swift-nio/blob/master/Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift#L314) data is  buffered until the upgrade is complete, and is than fired once the upgrade is done.

If you fire a `channel.write` immediatly after receiving one of those buffered `channel.read`s, NIO will error with 
```
The operation couldn’t be completed. (NIOHTTP1.NIOHTTPClientUpgradeError error 1.)
writingToHandlerAfterUpgradeCompleted
```
This is because the `write` will arrive in the `NIOHTTPClientUpgradeHandler` before it removed itself from the pipeline. The error is fired from [this line](https://github.com/apple/swift-nio/blob/master/Sources/NIOHTTP1/NIOHTTPClientUpgradeHandler.swift#L150) due to the line linked above setting `upgradeState` to `.upgradeComplete`.

### Expected behavior
My `channel.write` works fine, since `NIOHTTPClientUpgradeHandler` should remove itself from the pipeline

### Actual behavior
`NIOHTTPClientUpgradeHandler` receives the `write` before it removed itself from the pipeline.

### Steps to reproduce

1. Have a WebSocket server that sends back it's first packet with the HTTP upgrade response (this is important, otherwise data does not end up in the buffer)
2. Connect to the WebSocket
3. Fire a `channel.write` immediately after receiving the first packet.
4. Observe

### If possible, minimal yet complete reproducer code (or URL to code)
https://github.com/MrLotU/bug-repros/tree/nio-1279
All code outside of `main.swift` is copied from [vapor/websocket-kit](https://github.com/vapor/websocket-kit)

### SwiftNIO version/commit hash
SwiftNIO 2.10.1

### Swift & OS version (output of `swift --version && uname -a`)
```
Apple Swift version 5.1.2 (swiftlang-1100.0.278 clang-1100.0.33.9)
Target: x86_64-apple-darwin19.0.0
Darwin ip-192-168-178-248.eu-west-1.compute.internal 19.0.0 Darwin Kernel Version 19.0.0: Thu Oct 17 16:17:15 PDT 2019; root:xnu-6153.41.3~29/RELEASE_X86_64 x86_64
```
Also reproducible on latest Swift master toolchain (`org.swift.50201911301a`)

### PR DESCRIPTION
**Pull request by MrLotU**
Fixes #1279 

### Motivation:

Fixes the bug described in #1279 

### Modifications:

Moved state change in `NIOHTTPClientUpgradeHandler.performUpgrade` to after firing buffered messages.
 
### Result:

The `upgradeState` will now only change to `upgradeComplete` after buffered reads are handeld.

### CODE REVIEW

**Comment by MrLotU** [comment]
I'll give it a shot! 😄

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 1]
```diff
@@ -54,6 +54,39 @@ private func setUpClientChannel(clientHTTPHandler: RemovableChannelHandler,
     return channel
 }
 
+private final class ChannelReadWriteHandler: ChannelInboundHandler {
+    typealias InboundIn = Never
```
Comment (Lukasa) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: This should be `Any`.

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 2]
```diff
@@ -54,6 +54,39 @@ private func setUpClientChannel(clientHTTPHandler: RemovableChannelHandler,
...
+    typealias OutboundOut = Never
```
Comment (Lukasa) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: So should this.

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 3]
```diff
@@ -441,6 +474,42 @@ class HTTPClientUpgradeTestCase: XCTestCase {
...
+        XCTAssertNoThrow(try clientChannel.writeInbound(ByteBuffer.forString(response)))
+        
+        clientChannel.embeddedEventLoop.run()
+                
+        XCTAssert(upgradeHandlerCallbackFired)
+        
+        XCTAssertEqual(handler.messagesReceived, 1)
+        
+        XCTAssertNoThrow(try clientChannel.pipeline
+            .assertDoesNotContain(handlerType: NIOHTTPClientUpgradeHandler.self))
```
Comment (Lukasa) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Can you also do a `readByteBufferOutputAsString` and assert it contains the server response we wrote?

Comment (MrLotU) [reply Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: I tried adding this at the end, but the result is `nil`. However the `ByteBuffer` I get in my `ChannelReadWriteHandler` does contain the string I put in the response. Is there a step I'm skipping to have said buffer end up at the end. The pipeline looks like this:
```
ChannelPipeline[ObjectIdentifier(0x0000000103503a50)]:
                     [I] ↓↑ [O]
 ChannelReadWriteHandler ↓↑  [handler4]
```

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 4]
```diff
@@ -54,6 +54,39 @@ private func setUpClientChannel(clientHTTPHandler: RemovableChannelHandler,
...
+    
+    var messagesReceived = 0
+    
+    func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        self.messagesReceived += 1
+        context.writeAndFlush(data, promise: nil)
+    }
+}
+
+private final class AddHandlerClientUpgrader<T: ChannelInboundHandler>: NIOHTTPClientProtocolUpgrader {
```
Comment (weissi) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: can we move both those class definitions into the test itself?

**Comment by weissi** [comment]
@swift-nio-bot test this please

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift (hunk 3 above)
Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: You added `clientChannel.readByteBufferOutputAsString`? Whereabouts did you add that line?

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 5]
```diff
@@ -54,6 +54,39 @@ private func setUpClientChannel(clientHTTPHandler: RemovableChannelHandler,
     return channel
 }
 
+private final class ChannelReadWriteHandler: ChannelInboundHandler {
```
Comment (weissi) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: could you move these two `class`es into the new unit test function?

**Comment by weissi** [comment]
@MrLotU the new test is also missing:

```
11:31:31 => Checking linux tests... missing changes!
11:31:31 diff --git a/Tests/NIOHTTP1Tests/HTTPClientUpgradeTests+XCTest.swift b/Tests/NIOHTTP1Tests/HTTPClientUpgradeTests+XCTest.swift
11:31:31 index 616c9ba7..d5911d90 100644
11:31:31 --- a/Tests/NIOHTTP1Tests/HTTPClientUpgradeTests+XCTest.swift
11:31:31 +++ b/Tests/NIOHTTP1Tests/HTTPClientUpgradeTests+XCTest.swift
11:31:31 @@ -30,6 +30,7 @@ extension HTTPClientUpgradeTestCase {
11:31:31                  ("testSimpleUpgradeSucceeds", testSimpleUpgradeSucceeds),
11:31:31                  ("testUpgradeWithRequiredHeadersShowsInRequest", testUpgradeWithRequiredHeadersShowsInRequest),
11:31:31                  ("testSimpleUpgradeSucceedsWhenMultipleAvailableProtocols", testSimpleUpgradeSucceedsWhenMultipleAvailableProtocols),
11:31:31 +                ("testUpgradeCompleteFlush", testUpgradeCompleteFlush),
11:31:31                  ("testNoUpgradeAsNoServerUpgrade", testNoUpgradeAsNoServerUpgrade),
11:31:31                  ("testFirstResponseReturnsServerError", testFirstResponseReturnsServerError),
11:31:31                  ("testUpgradeResponseMissingAllProtocols", testUpgradeResponseMissingAllProtocols),
11:31:32 Build step 'Execute shell' marked build as failure
11:31:32 $ ssh-agent -k
```

please run `scripts/generate_linux_tests.rb` and record the changes it makes

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift (hunk 3 above)
Comment (MrLotU) [reply Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Right at the end. I can push the non-working version

Comment (Lukasa) [reply Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Please do.

Comment (MrLotU) [reply Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Pushed! 😄

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 6]
```diff
@@ -441,6 +441,77 @@ class HTTPClientUpgradeTestCase: XCTestCase {
...
+                                                    
+                                                    upgradeHandlerCallbackFired = true
+        }
+        defer {
+            XCTAssertNoThrow(try clientChannel.finish())
+        }
+        
+        // Read the server request.
+        if let requestString = try clientChannel.readByteBufferOutputAsString() {
+
```
Comment (Lukasa) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Delete this whitespace please.

File: Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift [hunk 7]
```diff
@@ -441,6 +441,77 @@ class HTTPClientUpgradeTestCase: XCTestCase {
...
+            XCTAssertNoThrow(try clientChannel.finish())
+        }
+        
+        // Read the server request.
+        if let requestString = try clientChannel.readByteBufferOutputAsString() {
+                        
+            XCTAssertEqual(requestString, "GET / HTTP/1.1\r\nContent-Type: text/plain; charset=utf-8\r\nContent-Length: 0\r\nConnection: upgrade\r\nUpgrade: myproto\r\n\r\n")
+        } else {
+            XCTFail()
+        }
```
Comment (Lukasa) [inline Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Under here add two more lines:

```
XCTAssertNoThrow(XCTAssertEqual(try clientChannel.readByteBufferOutputAsString(), ""))  // Empty body
XCTAssertNoThrow(XCTAssertNil(try clientChannel.readByteBufferOutputAsString()))
```

Then, please amend the code in `readByteBufferOutputAsString` in this file. Where we currently call `getString`, change it to `readString(length: requestBuffer.readableBytes)`, and change the binding of the buffer from `let` to `var`. Then things should start working.

Comment (MrLotU) [reply Tests/NIOHTTP1Tests/HTTPClientUpgradeTests.swift]: Awesome. Thanks!

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Pull request merged**